### PR TITLE
Introducing sampling mechanism and scope restoring (policy history resampling)

### PR DESCRIPTION
**Pull request by Danielhp95**

### CODE REVIEW

File: python/unitytrainers/self_play_trainer_controller.py [hunk 1]
```diff
@@ -16,8 +18,13 @@ def __init__(self, env_path, run_id, save_freq, curriculum_folder, fast_simulati
...
             os.makedirs(self.ghost_trainers_path)
-        raise UnityTrainerControllerException("The handle_episode_termination method was not implemented as hyper parameters need to be included")
-        # TODO add opci, delta and opponent save frequency hyperparameters
+
+        self.delta = None
+        self.opponent_policy_change_interval = None
+        self.ghost_save_frequency = None
+
+        self.elapsed_episodes = 0
+        raise UnityTrainerControllerException("The __init__ method was not implemented.")
```
Comment (Pete-York) [inline python/unitytrainers/self_play_trainer_controller.py:27]: What aspect is yet to be implemented?

File: python/unitytrainers/self_play_trainer_controller.py [hunk 2]
```diff
@@ -72,22 +79,26 @@ def save_models_and_checkpoints(self, sess, global_step, global_saver):
...
-    def restore_model_from_checkpoint(self, brain_name, checkpoint_path, sess):
-        self.ghost_trainer_savers[brain_name].restore(sess, checkpoint_path)
-
     def handle_episode_termination(self, curr_info, sess):
         # handle opponent policy change
         super(SelfPlayTrainerController, self).handle_episode_termination(curr_info, sess)
-        raise UnityTrainerControllerException("The handle_episode_termination method was not implemented.")
+        self.elapsed_episodes += 1
+        if self.should_change_ghost_model():
+            for brain_name, saver in self.ghost_trainers.items():
```
Comment (Pete-York) [inline python/unitytrainers/self_play_trainer_controller.py]: I would maybe refactor this loop into a method called resample_all_ghosts

**Review by Pete-York — COMMENTED** [review]
Mostly looks good, just a couple of comments

File: python/unitytrainers/self_play_trainer_controller.py (hunk 1 above)
Comment (gaoyuankidult) [reply python/unitytrainers/self_play_trainer_controller.py:27]: It might be better to specify what exactly needs to be implemented.

Comment (Danielhp95) [reply python/unitytrainers/self_play_trainer_controller.py:27]: What is left to be implemented is to specify where the self play hyper parameters are introduced. Most likely in the `trainer_config.yaml`?

File: python/unitytrainers/self_play_trainer_controller.py (hunk 2 above)
Comment (Danielhp95) [reply python/unitytrainers/self_play_trainer_controller.py]: Sure, will do

**Pull request merged**